### ISSUE TITLE
"no-unused-prop-types" fires incorrectly, if this.props is passed through a default function parameter.

### ISSUE THREAD

**Issue by abenitesDC** [yannickcr/eslint-plugin-react#1658]
Introduced in `v7.6.0`. `eslint-plugin-react` incorrectly fires `no-used-prop-types` rule, if props is used as a default function parameter. This was working correctly on 7.5.1.

```javascript
const myComponent = createReactClass({
  propTypes: {
    foo: PropTypes.string // throws incorrectly eslint rule: 'foo' PropType is defined but prop is never used (react/no-unused-prop-types)
  },

  boo: function(props = this.props) {
    const {foo} = props;

    return foo;
  }
});
```

**Comment by ljharb** [comment]
This is actually correct; the linter has no way of knowing if `boo` is called with the actual props object, or some random object.

The proper solution is to destructure `foo` at the callsite, and pass *that* into `boo`.
